File: packages/core-api/src/compression.ts
import { gzipSync, gunzipSync } from 'node:zlib'
import { DOM_DUMP_COMPRESSION_THRESHOLD, CompressedDomDump, DomNode, DomDump } from '../schemas/DomNode.schema'

/**
 * Checks if the given data is in compressed DomDump format
 */
export function isCompressedDomDump(data: unknown): data is CompressedDomDump {
    return (
        data !== null &&
        typeof data === 'object' &&
        'compressed' in data &&
        (data as CompressedDomDump).compressed === true &&
        'data' in data &&
        typeof (data as CompressedDomDump).data === 'string'
    )
}

/**
 * Compresses DomDump if it exceeds the threshold size
 * Returns original string if below threshold, or CompressedDomDump if compressed
 *
 * @param domDump - DomNode tree or JSON string
 * @returns Original string if small, or CompressedDomDump if large
 */
export function compressDomDump(domDump: DomNode | string): string | CompressedDomDump {
    const jsonString = typeof domDump === 'string'
        ? domDump
        : JSON.stringify(domDump)

    const originalSize = Buffer.byteLength(jsonString, 'utf8')

    // Skip compression for small payloads
    if (originalSize <= DOM_DUMP_COMPRESSION_THRESHOLD) {
        return jsonString
    }

    const compressed = gzipSync(Buffer.from(jsonString, 'utf8'))

    return {
        data: compressed.toString('base64'),
        compressed: true,
        originalSize,
    }
}

/**
 * Decompresses DomDump if it was compressed
 * Handles both compressed and uncompressed formats
 *
 * @param data - String, CompressedDomDump, or DomNode object
 * @returns Parsed DomNode object or null if invalid
 */
export function decompressDomDump(data: string | CompressedDomDump | DomNode): DomNode | null {
    try {
        // Already a DomNode object
        if (typeof data === 'object' && !isCompressedDomDump(data)) {
            return data as DomNode
        }

        // Compressed format
        if (isCompressedDomDump(data)) {
            const buffer = Buffer.from(data.data, 'base64')
            const decompressed = gunzipSync(buffer).toString('utf8')
            return JSON.parse(decompressed) as DomNode
        }

        // JSON string
        if (typeof data === 'string') {
            // Try to parse - might be a JSON string of compressed data
            const parsed = JSON.parse(data)

            // Check if the parsed result is compressed format
            if (isCompressedDomDump(parsed)) {
                return decompressDomDump(parsed)
            }

            return parsed as DomNode
        }

        return null
    } catch (e) {
        console.error('Failed to decompress domDump:', e)
        return null
    }
}

/**
 * Prepares DomDump for sending to the server
 * Compresses if necessary and returns the appropriate format
 *
 * @param domDump - DomNode tree, string, or already compressed data
 * @returns Object with processed data and compression flag
 */
export function prepareDomDumpForTransfer(domDump: DomDump | string): {
    data: string;
    isCompressed: boolean;
} {
    // If already compressed, just stringify it
    if (isCompressedDomDump(domDump)) {
        return {
            data: JSON.stringify(domDump),
            isCompressed: true,
        }
    }

    // Otherwise, try to compress
    const result = compressDomDump(domDump)

    if (isCompressedDomDump(result)) {
        return {
            data: JSON.stringify(result),
            isCompressed: true,
        }
    }

    return {
        data: result,
        isCompressed: false,
    }
}

File: packages/core-api-python/syngrisi_core_api/compression.py
"""DOM dump compression utilities ported from JS `src/compression.ts`.

Uses gzip + base64. The compression threshold is 50KB (DOM_DUMP_COMPRESSION_THRESHOLD).
"""

import base64
import gzip
import json

from .schemas import DOM_DUMP_COMPRESSION_THRESHOLD


def _json_stringify(value):
    """Equivalent of JS JSON.stringify (compact, no extra whitespace)."""
    return json.dumps(value, separators=(",", ":"), ensure_ascii=False)


def is_compressed_dom_dump(data):
    """True iff ``data`` is a dict with ``compressed == True`` and a string ``data``."""
    return (
        isinstance(data, dict)
        and data.get("compressed") is True
        and isinstance(data.get("data"), str)
    )


def compress_dom_dump(dom_dump):
    """Compress a DOM dump if it exceeds the threshold.

    Returns the original JSON string when below the threshold, otherwise a
    ``{data, compressed: True, originalSize}`` dict.
    """
    json_string = dom_dump if isinstance(dom_dump, str) else _json_stringify(dom_dump)

    original_size = len(json_string.encode("utf-8"))

    # Skip compression for small payloads.
    if original_size <= DOM_DUMP_COMPRESSION_THRESHOLD:
        return json_string

    compressed = gzip.compress(json_string.encode("utf-8"))

    return {
        "data": base64.b64encode(compressed).decode("ascii"),
        "compressed": True,
        "originalSize": original_size,
    }


def decompress_dom_dump(data):
    """Decompress a DOM dump, handling compressed/uncompressed/string formats.

    Returns the parsed DomNode object or ``None`` if invalid.
    """
    try:
        # Already a DomNode object (dict, not compressed).
        if isinstance(data, dict) and not is_compressed_dom_dump(data):
            return data

        # Compressed format.
        if is_compressed_dom_dump(data):
            buffer = base64.b64decode(data["data"])
            decompressed = gzip.decompress(buffer).decode("utf-8")
            return json.loads(decompressed)

        # JSON string.
        if isinstance(data, str):
            parsed = json.loads(data)
            if is_compressed_dom_dump(parsed):
                return decompress_dom_dump(parsed)
            return parsed

        return None
    except Exception as e:  # noqa: BLE001 - mirror JS catch-all
        print(f"Failed to decompress domDump: {e}")
        return None


def prepare_dom_dump_for_transfer(dom_dump):
    """Prepare a DOM dump for transfer.

    Returns ``{"data": str, "isCompressed": bool}``.
    """
    # If already compressed, just stringify it.
    if is_compressed_dom_dump(dom_dump):
        return {"data": _json_stringify(dom_dump), "isCompressed": True}

    result = compress_dom_dump(dom_dump)

    if is_compressed_dom_dump(result):
        return {"data": _json_stringify(result), "isCompressed": True}

    return {"data": result, "isCompressed": False}


# JS-style aliases.
isCompressedDomDump = is_compressed_dom_dump
compressDomDump = compress_dom_dump
decompressDomDump = decompress_dom_dump
prepareDomDumpForTransfer = prepare_dom_dump_for_transfer
